### PR TITLE
feat(stepfunctions): added new condition operators 

### PR DESCRIPTION
**Pull request by michaelwiles**
Added the following Conditons:
* `Condition.isPresent` - matches if a json path is present
* `Condition.isNotPresent` - matches if a json path is not present
* `Condition.isString` - matches if a json path contains a string
* `Condition.isNotString` - matches if a json path is not a string
* `Condition.isNumeric` - matches if a json path is numeric
* `Condition.isNotNumeric` - matches if a json path is not numeric
* `Condition.isBoolean` - matches if a json path is boolean
* `Condition.isNotBoolean` - matches if a json path is not boolean
* `Condition.isTimestamp` - matches if a json path is a timestamp
* `Condition.isNotTimestamp` - matches if a json path is not a
timestamp
* `Condition.isNotNull` - matches if a json path is not null
* `Condition.isNull` - matches  if a json path is null
* `Condition.booleanEqualsJsonPath` - matches if a boolean field equals
a value in a given mapping path
* `Condition.stringEqualsJsonPath` - matches if a string field equals a
given mapping path
* `Condition.stringLessThanJsonPath` - Matches if a string field sorts
before a value at given mapping path
* `Condition.stringLessThanEqualsJsonPath` - Matches if a string field
sorts equal to or before a given mapping
* `Condition.stringGreaterThanJsonPath` - Matches if a string field
sorts after a value at a given mapping path
* `Condition.stringGreaterThanEqualsJsonPath` - Matches if a string
field sorts after or equal to value at a given mapping path
* `Condition.numberEqualsJsonPath` - matches if a numeric field has the
value in a given mapping path
* `Condition.numberLessThan` - matches if a numeric field is less than
the given value
* `Condition.numberLessThanJsonPath` - matches if a numeric field is
less than the value at the given mapping path
* `Condition.numberLessThanEqualsJsonPath` - matches if a numeric field
is less than or equal to the numeric value at given mapping path
* `Condition.numberGreaterThanJsonPath` - matches if a numeric field is
greater than the value at a given mapping path
* `Condition.numberGreaterThanEqualsJsonPath` - matches if a numeric
field is greater than or equal to the value at a given mapping path
* `Condition.timestampEqualsJsonPath` - matches if a timestamp field is
the same time as the timestamp at a given mapping path
* `Condition.timestampLessThanJsonPath` - matches if a timestamp field
is before the timestamp at a given mapping path
* `Condition.timestampLessThanEqualsJsonPath` - matches if a timestamp
field is before or equal to the timestamp at a given mapping path
* `Condition.timestampGreaterThanJsonPath` - matches if a timestamp
field is after the timestamp at a given mapping path
* `Condition.timestampGreaterThanEqualsJsonPath` - matches if a
timestamp field is after or equal to the timestamp at a given mapping
path
* `Condition.stringMatches` - matches a field with the ability to use as
a wild card e.g: log-.txt or LATEST. No other characters other than "*"
have any special meaning (* can be escaped: \\*)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-stepfunctions/lib/condition.ts [hunk 1]
```diff
@@ -2,13 +2,110 @@
...
+  /**
+   * Tests if variable is a timestamp
+   */
+  public static isTimestamp(variable: string): Condition {
+    return new VariableComparison(variable, ComparisonOperator.IsTimestamp, true);
+  }
+
+  /**
+   * Tests if variable is not a timestamp
+   */
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions/lib/condition.ts:71]: the terminology `tests` and `matches` are used. can we be consistent (I think I prefer Matches which already exists on a the comparison operators that are available today)

File: packages/@aws-cdk/aws-stepfunctions/lib/condition.ts [hunk 2]
```diff
@@ -2,13 +2,110 @@
...
+  }
   /**
    * Matches if a boolean field has the given value
    */
   public static booleanEquals(variable: string, value: boolean): Condition {
     return new VariableComparison(variable, ComparisonOperator.BooleanEquals, value);
   }
 
+  /**
+   * Matches if a boolean field has the given value
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions/lib/condition.ts:96]:
```suggestion
   * Matches if a boolean field has the given value path
```

File: packages/@aws-cdk/aws-stepfunctions/lib/condition.ts [hunk 3]
```diff
@@ -2,13 +2,110 @@
...
+   */
+  public static booleanEqualsPath(variable: string, value: string): Condition {
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions/lib/condition.ts:98]: wondering if `Path` is the right suffix for these methods?
I think we've been explicit about adding `fromJsonPath` when naming these methods. what do you think?

File: packages/@aws-cdk/aws-stepfunctions/lib/condition.ts [hunk 4]
```diff
@@ -51,69 +176,146 @@ export abstract class Condition {
...
+
   /**
    * Matches if a numeric field is less than or equal to the given value
    */
   public static numberLessThanEquals(variable: string, value: number): Condition {
     return new VariableComparison(variable, ComparisonOperator.NumericLessThanEquals, value);
   }
 
+  /**
+   * Matches if a numeric field is less than or equal to the numeric value at given maping path
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions/lib/condition.ts:208]:
```suggestion
   * Matches if a numeric field is less than or equal to the numeric value at given mapping path
```

File: packages/@aws-cdk/aws-stepfunctions/test/condition.test.ts [hunk 5]
```diff
@@ -28,13 +28,85 @@ describe('Condition Variables', () => {
       },
     );
   }),
-  test('Exercise a number of other conditions', () => {
+  test('Exercise string conditions', () => {
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions/test/condition.test.ts:31]: ideally we do not modify existing tests. it's preferred to only add new tests to ensure we don't introduce a regression. can we preserve this test as it was and test our conditions through new tests cases only?

**Review by shivlaks — CHANGES_REQUESTED** [review]
thanks for starting this @michaelwiles 

- [ ] The commit body should be updated with the list of condition operators we are adding support for.
- [ ] can you update the `README` as well to include the new conditions
- [ ] had a question about naming the methods that will support a JsonPath

File: packages/@aws-cdk/aws-stepfunctions/lib/condition.ts (hunk 1 above)
Comment (michaelwiles) [reply packages/@aws-cdk/aws-stepfunctions/lib/condition.ts:71]: happy with that

File: packages/@aws-cdk/aws-stepfunctions/lib/condition.ts (hunk 2 above)
Comment (michaelwiles) [reply packages/@aws-cdk/aws-stepfunctions/lib/condition.ts:96]: will fix

File: packages/@aws-cdk/aws-stepfunctions/lib/condition.ts (hunk 3 above)
Comment (michaelwiles) [reply packages/@aws-cdk/aws-stepfunctions/lib/condition.ts:98]: Yeah - you have a point. In the Dynamo attribute value context json path is used.

I guess in this case I took inspiration from the cloudformation naming which doesn't use the json path. 

I'm happy to stick with json path... don't have strong feelings against it

**Comment by shivlaks** [comment]
@yoodan93 @wong-a - can y'all have a look at this PR as well?

**Comment by michaelwiles** [comment]
@shivlaks changes implemented

File: packages/@aws-cdk/aws-stepfunctions/README.md [hunk 6]
```diff
@@ -217,6 +217,53 @@ If your `Choice` doesn't have an `otherwise()` and none of the conditions match
...
+* `Condition.timestampLessThan` - Matches if a timestamp field is before the given timestamp
+* `Condition.timestampLessThanJsonPath` - Matches if a timestamp field is before the timestamp at a given mapping path
+* `Condition.timestampLessThanEquals` - Matches if a timestamp field is before or equal to the given timestamp
+* `Condition.timestampLessThanEqualsJsonPath` - Matches if a timestamp field is before or equal to the timestamp at a given mapping path
+* `Condition.timestampGreaterThan` - Matches if a timestamp field is after the timestamp at a given mapping path
+* `Condition.timestampGreaterThanJsonPath` - Matches if a timestamp field is after the timestamp at a given mapping path
+* `Condition.timestampGreaterThanEquals` - Matches if a timestamp field is after or equal to the given timestamp
+* `Condition.timestampGreaterThanEqualsJsonPath` - Matches if a timestamp field is after or equal to the timestamp at a given mapping path
+* `Condition.stringMatches` 
 ### Parallel
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions/README.md:267]:
```suggestion
* `Condition.stringMatches` 

### Parallel
```

**Review by shivlaks — APPROVED** [review]
@michaelwiles - changes look fine to me! can you update the commit body to include the implementation details / documentation, etc for this commit?

note: also added a `do-not-merge` label so that @yoodan93 and @wong-a have a chance to take a look at the changes as they might have additional suggestions.

File: packages/@aws-cdk/aws-stepfunctions/README.md (hunk 6 above)
Comment (michaelwiles) [reply packages/@aws-cdk/aws-stepfunctions/README.md]: changed existing commit to incorporate this

**Comment by michaelwiles** [comment]
commit body updated

**Review by yoodan93 — APPROVED** [review]
Looks good to me!

**Comment by shivlaks** [comment]
> commit body updated

@michaelwiles by commit body i meant the description of this PR :) 
sorry, i should have been clearer!!! - the PR description will be used as the commit body when merging

**Review by wong-a — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-stepfunctions/README.md [hunk 7]
```diff
@@ -217,6 +217,54 @@ If your `Choice` doesn't have an `otherwise()` and none of the conditions match
 the JSON state, a `NoChoiceMatched` error will be thrown. Wrap the state machine
 in a `Parallel` state if you want to catch and recover from this.
 
+#### Available Conditions: 
+see [step function comparison operators](https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-choice-state.html#amazon-states-language-choice-state-rules)
+* `Condition.isJsonPathPresent` - matches a json path is present
```
[the remark this answers is not in the document]
Comment (michaelwiles) [reply packages/@aws-cdk/aws-stepfunctions/README.md:222]: not sure what you are asking

[the remark this answers is not in the document]
Comment (wong-a) [reply packages/@aws-cdk/aws-stepfunctions/README.md:222]: I meant that `isJsonPathPresent`, `isJsonPathNotPresent`, etc. aren't defined in  the `Condition` class

[the remark this answers is not in the document]
Comment (michaelwiles) [reply packages/@aws-cdk/aws-stepfunctions/README.md:222]: Very good point - thank you. The jsonpath here is superfluous and does not match the code - will address.

[the remark this answers is not in the document]
Comment (michaelwiles) [reply packages/@aws-cdk/aws-stepfunctions/README.md]: updated accordingly - made README match the code.

File: packages/@aws-cdk/aws-stepfunctions/README.md [hunk 8]
```diff
@@ -217,6 +217,54 @@ If your `Choice` doesn't have an `otherwise()` and none of the conditions match
...
+* `Condition.isJsonPathNotString` - matches a json path is not a string
+* `Condition.isJsonPathNumeric` - matches a json path is numeric
+* `Condition.isJsonPathNotNumeric` - matches a json path is not numeric
+* `Condition.isJsonPathBoolean` - matches a json path is boolean
+* `Condition.isJsonPathNotBoolean` - matches a json path is not boolean
+* `Condition.isJsonPathTimestamp` - matches a json path is a timestamp
+* `Condition.isJsonPathNotTimestamp` - matches a json path is not a timestamp
+* `Condition.isJsonPathNotNull` - matches a json path is not null
+* `Condition.isJsonPathNull` - matches a json path is null
+* `Condition.booleanEquals` - matches a boolean field has a given value
```
[the remark this answers is not in the document]
Comment (michaelwiles) [reply packages/@aws-cdk/aws-stepfunctions/README.md]: and fixed consistency of case and wording

**Review by wong-a — APPROVED** [review]

**Comment by michaelwiles** [comment]
@shivlaks please can you move this along? we want to use these new conditions...

**Comment by shivlaks** [comment]
@michaelwiles thanks for the contribution! dropped the `do-not-merge` label. This feature will be included in the next release.

**Comment by michaelwiles** [comment]
@shivlaks it looks like you still need to approve the pull request..?

**Review by shivlaks — APPROVED** [review]

**Pull request merged**